src/main/telemetry/crsf.c

```diff
@@ -326,6 +359,80 @@ void crsfFrameDeviceInfo(sbuf_t *dst) {
     *lengthPtr = sbufPtr(dst) - lengthPtr;
 }
 
+
+#if defined(USE_CRSF_V3)
+void crsfFrameSpeedNegotiationResponse(sbuf_t *dst, bool reply)
+{
+
+    uint8_t *lengthPtr = sbufPtr(dst);
+    sbufWriteU8(dst, 0);
+    sbufWriteU8(dst, CRSF_FRAMETYPE_COMMAND);
+    sbufWriteU8(dst, CRSF_ADDRESS_CRSF_RECEIVER);
+    sbufWriteU8(dst, CRSF_ADDRESS_FLIGHT_CONTROLLER);
+    sbufWriteU8(dst, CRSF_COMMAND_SUBCMD_GENERAL);
+    sbufWriteU8(dst, CRSF_COMMAND_SUBCMD_GENERAL_CRSF_SPEED_RESPONSE);
+    sbufWriteU8(dst, crsfSpeed.portID);
+    sbufWriteU8(dst, reply);
+    crc8_poly_0xba_sbuf_append(dst, &lengthPtr[1]);
+    *lengthPtr = sbufPtr(dst) - lengthPtr;
+}
+
+static void crsfProcessSpeedNegotiationCmd(uint8_t *frameStart)
+{
+
+    uint32_t newBaudrate = frameStart[2] << 24 | frameStart[3] << 16 | frameStart[4] << 8 | frameStart[5];
+    uint8_t ii = 0;
+    for (ii = 0; ii < BAUD_COUNT; ++ii) {
+        if (newBaudrate == baudRates[ii]) {
+            break;
+        }
+    }
+    crsfSpeed.portID = frameStart[1];
+    crsfSpeed.index = ii;
+}
+
+void crsfScheduleSpeedNegotiationResponse(void)
+{
+    crsfSpeed.hasPendingReply = true;
+    crsfSpeed.isNewSpeedValid = false;
+}
+
+void speedNegotiationProcess(uint32_t currentTime)
+{
+    if (crsfSpeed.hasPendingReply) {
+        bool found = crsfSpeed.index < BAUD_COUNT ? true : false;
+        sbuf_t crsfSpeedNegotiationBuf;
+        sbuf_t *dst = &crsfSpeedNegotiationBuf;
+        crsfInitializeFrame(dst);
+        crsfFrameSpeedNegotiationResponse(dst, found);
+        crsfFinalize(dst);
+        crsfRxSendTelemetryData();
+        crsfSpeed.hasPendingReply = false;
+        crsfSpeed.isNewSpeedValid = true;
+        crsfSpeed.confirmationTime = currentTime;
+        return;
+    } else if (crsfSpeed.isNewSpeedValid) {
+        if (currentTime - crsfSpeed.confirmationTime >= 4000) {
+            // delay 4ms before applying the new baudrate
+            crsfRxUpdateBaudrate(getCrsfDesireSpeed());
+            crsfSpeed.isNewSpeedValid = false;
+            isCrsfV3Running = true;
+            return;
+        }
+    }
+
+    // to notify the RX to fall back by sedning device info frame if telemetry is disabled
+    if (!featureIsEnabled(FEATURE_TELEMETRY) && getCrsfDesireSpeed() == CRSF_BAUDRATE) {
```

Comment:
The early exits above and this condition seem to result in a weird mix of this being checked or not, depending on control exiting early - should this be checked at the top, and the rest of the code be inside the `else`?